### PR TITLE
ARROW-16546: [Parquet][C++][Python] Make Thrift limits configurable

### PR DESCRIPTION
**Pull request by pitrou**
In fringe cases, users may have Parquet files where deserializing exceeds our default Thrift size limits.

### CODE REVIEW

File: cpp/src/parquet/metadata.h [hunk 1]
```diff
@@ -126,6 +126,12 @@ class PARQUET_EXPORT ColumnChunkMetaData {
       const ApplicationVersion* writer_version = NULLPTR, int16_t row_group_ordinal = -1,
       int16_t column_ordinal = -1,
       std::shared_ptr<InternalFileDecryptor> file_decryptor = NULLPTR);
```
[the remark this answers is not in the document]
Comment (lidavidm) [reply cpp/src/parquet/metadata.h:128]: I think we can just keep it. Maybe put a deprecation on it?

File: python/pyarrow/_dataset_parquet.pyx [hunk 2]
```diff
@@ -654,17 +668,49 @@ cdef class ParquetFragmentScanOptions(FragmentScanOptions):
...
+        self.reader_properties().set_thrift_string_size_limit(size)
+
+    @property
+    def thrift_container_size_limit(self):
+        return self.reader_properties().thrift_container_size_limit()
+
+    @thrift_container_size_limit.setter
+    def thrift_container_size_limit(self, size):
+        if size <= 0:
+            raise ValueError("size must be larger than zero")
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset_parquet.pyx:688]: Should such a check be included in `_parquet.pyx` as well?

File: python/pyarrow/parquet/__init__.py [hunk 3]
```diff
@@ -2258,11 +2269,13 @@ class _ParquetDatasetV2:
...
     """
 
-    def __init__(self, path_or_paths, filesystem=None, filters=None,
+    def __init__(self, path_or_paths, filesystem=None, *, filters=None,
                  partitioning="hive", read_dictionary=None, buffer_size=None,
                  memory_map=False, ignore_prefixes=None, pre_buffer=True,
                  coerce_int96_timestamp_unit=None, schema=None,
-                 decryption_properties=None, **kwargs):
+                 decryption_properties=None, thrift_string_size_limit=None,
+                 thrift_container_size_limit=None,
```
Comment (jorisvandenbossche) [inline python/pyarrow/parquet/__init__.py:2277]: I think you need to add it to the signature of `class ParquetDataset.__new__` as well, to pass it through

Comment (jorisvandenbossche) [reply python/pyarrow/parquet/__init__.py:2277]: Although if the tests pass ..

File: python/pyarrow/_dataset_parquet.pyx (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/_dataset_parquet.pyx:688]: Will do.

File: python/pyarrow/parquet/__init__.py (hunk 3 above)
Comment (pitrou) [reply python/pyarrow/parquet/__init__.py:2277]: Yeah, `pq.read_table` invokes dataset directly.

Comment (jorisvandenbossche) [reply python/pyarrow/parquet/__init__.py:2283]: Ah, yes, I see

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

File: cpp/src/parquet/column_reader.cc [hunk 4]
```diff
@@ -223,14 +223,15 @@ EncodedStatistics ExtractStatsFromHeader(const H& header) {
 class SerializedPageReader : public PageReader {
  public:
   SerializedPageReader(std::shared_ptr<ArrowInputStream> stream, int64_t total_num_rows,
-                       Compression::type codec, ::arrow::MemoryPool* pool,
+                       Compression::type codec, const ReaderProperties& properties,
```
Comment (emkornfield) [inline cpp/src/parquet/column_reader.cc:226]: is properties a movable type?  should we pass by value and move?

File: cpp/src/parquet/metadata.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (emkornfield) [reply cpp/src/parquet/metadata.h:130]: +1 to deprecation.

File: cpp/src/parquet/thrift_internal.h [hunk 5]
```diff
@@ -350,74 +351,84 @@ static inline format::EncryptionAlgorithm ToThrift(EncryptionAlgorithm encryptio
...
+        container_size_limit_(container_size_limit) {}
 
-inline std::shared_ptr<ThriftBuffer> CreateReadOnlyMemoryBuffer(uint8_t* buf,
-                                                                uint32_t len) {
+  // Deserialize a thrift message from buf/len.  buf/len must at least contain
+  // all the bytes needed to store the thrift message.  On return, len will be
+  // set to the actual length of the header.
+  template <class T>
+  void DeserializeMessage(const uint8_t* buf, uint32_t* len, T* deserialized_msg,
+                          const std::shared_ptr<Decryptor>& decryptor = NULLPTR) {
```
Comment (emkornfield) [inline cpp/src/parquet/thrift_internal.h:369]: why shared_ptr here?  If this is needed for async operations doesn't a copy of the shared_ptr need to be taken?  If it isn't pass by normal pointer?

File: python/pyarrow/_dataset_parquet.pyx [hunk 6]
```diff
@@ -654,17 +668,49 @@ cdef class ParquetFragmentScanOptions(FragmentScanOptions):
     def pre_buffer(self, bint pre_buffer):
         self.arrow_reader_properties().set_pre_buffer(pre_buffer)
 
+    @property
+    def thrift_string_size_limit(self):
+        return self.reader_properties().thrift_string_size_limit()
+
+    @thrift_string_size_limit.setter
+    def thrift_string_size_limit(self, size):
+        if size <= 0:
```
Comment (emkornfield) [inline python/pyarrow/_dataset_parquet.pyx:677]: should these checks be pushed into C++?

File: cpp/src/parquet/column_reader.cc (hunk 4 above)
Comment (pitrou) [reply cpp/src/parquet/column_reader.cc:226]: Yes, we probably can.

File: cpp/src/parquet/thrift_internal.h (hunk 5 above)
Comment (emkornfield) [reply cpp/src/parquet/thrift_internal.h:369]: I see this is just moving around code.

Comment (pitrou) [reply cpp/src/parquet/thrift_internal.h:369]: It is just the original signature in the previous code, but I can change it.

File: python/pyarrow/_dataset_parquet.pyx (hunk 6 above)
Comment (pitrou) [reply python/pyarrow/_dataset_parquet.pyx:677]: We don't seem to validate Parquet properties values on the C++ side.

**Review by emkornfield — APPROVED** [review]
Took a quick pass through.  LGTM, any plans to add a sample file with overrides that would pass/fail to verify end-to-end functionality (sorry if I missed this).

**Comment by pitrou** [comment]
> any plans to add a sample file with overrides that would pass/fail to verify end-to-end functionality (sorry if I missed this)

The Python tests generate such a file on the fly.

**Pull request merged**